File: w10-mean/d3-drills-dfs/rb/tree.rb
##
# This class represents a tree
# where each node has an array of children.

class Tree

  ##
  # reads the key of the tree's root
  attr_reader :key

  ## 
  # sets the key of the tree's root node
  # unless intended new key is nil
  # (to remove a node, set the entire node to nil)
  def key=(new_key)
    if !new_key.nil?
      @key = new_key
    end
    @key
  end

  ##
  # sets or reads the array of children for this instance
  attr_accessor :children


  ##
  # Creates a new tree with the root node key specified by 
  # the +key+ param as the @key for the instance
  def initialize(key)
    @key = key
    @children = []
  end

  ##
  # Searches through all nodes of the tree, spreading 
  # outward from the root. Looks for any node with key equal
  # to the +target_key+ param. Returns nil if no such node is found.
  def breadth_first_search(target_key)
    nil
  end

  ##
  # Searches through tree for a single node whose key is equal to
  # the +target_key+ param. Looks through nodes in a depth-first order,
  # by following each path as deep as it can before backtracking.
  # Returns nil if no such node is found.
  def depth_first_search(target_key)
    nil
  end

end


my_tree = Tree.new('Q')
r_node = Tree.new('R')
s_node = Tree.new('S')
t_node = Tree.new('T')
u_node = Tree.new('U')
v_node = Tree.new('V')
w_node = Tree.new('W')

my_tree.children << r_node
my_tree.children << s_node
r_node.children << t_node
r_node.children << u_node
r_node.children << v_node
u_node.children << w_node

#          T 
#        /
#     R  -- U -- W
#   /   \
# Q        V
#   \
#     S 

puts "\n-- expect node U --"
u_result = my_tree.depth_first_search('U')
p u_result


puts "\n-- expect node W --"
w_result = my_tree.depth_first_search('W')
p w_result


puts "\n-- expect nil -- "
z_result = my_tree.depth_first_search('Z')
p z_result

File: w10-mean/d3-drills-dfs/js/tree.js
// Note: To run only recursive version of depth first search tests from the Terminal 
// use `node tree-solution.js recursive`.
// For iterative only, use `node tree-solution.js iterative`.

// This object type represents a tree
// where each node has an array of children.

var TreeNode = function (key, children){
	this.key = key; 	
	this.children = children || [];	// an array, holds the Trees that we can go to directly from here
};

// Searches through tree for a single node whose key is equal to
// the +target_key+ param. Looks through nodes in a depth-first order,
// by following each path as deep as it can before backtracking.
// Returns null if no such node is found.
TreeNode.prototype.depthFirstSearch	= function(targetKey) {
	return null;
};

// Searches through all nodes of the tree, spreading 
// outward from the root. Looks for any node with key equal
// to the targetKey param. Returns null if no such node is found.
TreeNode.prototype.breadthFirstSearch = function(targetKey) {
	return null;
};


// Returns an array containing all nodes for which 
// the selectionFunction returns true.
TreeNode.prototype.breadthFirstSearchAll = function(selectionFunction){
	return [];
};


var myTree = new TreeNode('Q'),
	rNode = new TreeNode('R'),
	sNode = new TreeNode('S'),
	tNode = new TreeNode('T'),
	uNode = new TreeNode('U'),
	vNode = new TreeNode('V'),
	wNode = new TreeNode('W');

myTree.children.push(rNode, sNode);
rNode.children.push(tNode, uNode, vNode);
uNode.children.push(wNode);

//          T 
//        /
//     R  -- U -- W
//   /	  \
// Q        V
//   \
//     S 


console.log("\n-- expect node U --");
console.log(myTree.depthFirstSearch('U'));  



console.log("\n-- expect node W --");
console.log(myTree.depthFirstSearch('W'));  



console.log("\n-- expect null --");
console.log(myTree.depthFirstSearch('Z')); 



